hail/src/main/scala/is/hail/shuffler/ShuffleClient.scala

```diff
@@ -1,99 +1,165 @@
 package is.hail.shuffler
 
-import java.io._
+import java.net.Socket
 
+import is.hail._
 import is.hail.annotations._
-import is.hail.expr.ir.ExecuteContext
-import is.hail.types.virtual.TStruct
-import is.hail.io.TypedCodecSpec
+import is.hail.expr.ir._
+import is.hail.types.encoded._
+import is.hail.types.virtual._
 import is.hail.utils._
 import javax.net.ssl._
 import org.apache.log4j.Logger
 
+object ShuffleClient {
+  private[this] val log = Logger.getLogger(getClass.getName())
```

Comment:
I agree, the Right Thing seems to be a static class variable. I aesthetically dislike adding `import ShuffleClient._` to the top of my `class`es to get static variables in scope from the `object`. That's why I stick `private[this] val log = ...` in the `class`.

If the concern is allocation of logger objects, you only get one instance per key. If you follow the levels of indirection about five classes deep you get to: [`Hierarchy.getLogger`](https://github.com/apache/log4j/blob/7be00eed88152dd011a619e8bae5a631235c3f4c/src/main/java/org/apache/log4j/Hierarchy.java). That guy has a map from keys to loggers.

That said, if you still prefer this approach, I'm happy to make the changes you've requested in another PR. It looks like we already have `trait Logging` in utils, but it uses the root logger for everything. That obscures class source, which I strongly dislike. In this hypothetical PR, I'll have to think about whether we want to fix everything to use `extends Logging` rather than using the root logger.

Reply:
> If the concern is allocation of logger objects

It is not.

I missed that you also added the `private[this] val log = ...` to the class (although you have to since you made it `private[this]`.)

You prefer this to `import ShuffleClient._` when this is (1) longer and (2) adds a logger lookup on construction, and a local variable to the class?  That's fine for something like the shuffle client, but doesn't seem great for lightweight classes that but nonetheless want to log.

I didn't mean our actual `Logger` class, which is clearly garbage.

I don't actually know what we should do here.  Some options:
 - yours (every class does  `private[this] val log = ...`)
 - a mixing trait that has `protected val log = ...` and every class that wants to log mixes it in.
 - Mine: `private val log = ...` on the object (or via a mixing) and `import Object._` in the class.
 - A logger on the package object.  It will be visible everywhere without an import.  But you lose class information.

The last one almost works, you can actually get the class via introspection but it is extremely slow.

So leave it for now, but I used my suggestion (with the explicit name without the mangling and no mixing) so we've got a bit of divergence going on.